sdks/python/apache_beam/runners/worker/bundle_processor.py

```diff
@@ -241,22 +252,54 @@ def process_bundle(self, instruction_id):
         # We must wait until we receive "end of stream" for each of these ops.
         expected_inputs.append(op)
 
-    # Start all operations.
-    for op in reversed(ops):
-      logging.info('start %s', op)
-      op.start()
-
-    # Inject inputs from data plane.
-    for input_op in expected_inputs:
-      for data in input_op.data_channel.input_elements(
-          instruction_id, [input_op.target]):
-        # ignores input name
-        input_op.process_encoded(data.data)
-
-    # Finish all operations.
-    for op in ops:
-      logging.info('finish %s', op)
-      op.finish()
+    try:
+      self.state_sampler.start()
+      # Start all operations.
+      for op in reversed(self.ops.values()):
+        logging.info('start %s', op)
+        op.start()
+
+      # Inject inputs from data plane.
+      for input_op in expected_inputs:
+        for data in input_op.data_channel.input_elements(
+            instruction_id, [input_op.target]):
+          # ignores input name
+          input_op.process_encoded(data.data)
+
+      # Finish all operations.
+      for op in self.ops.values():
+        logging.info('finish %s', op)
+        op.finish()
+    finally:
+      self.state_sampler.stop_if_still_running()
+
+  def metrics(self):
+    return beam_fn_api_pb2.Metrics(
+        # TODO(robertwb): Rename to progress?
```

Comment:
Are you thinking about renaming the Metrics object or the ptransforms field to progress?
Also, I'm pretty happy with the naming structure of Metrics, what made you think that it should change?

Reply:
I would name the transforms field to progress. (And possibly Metrics.PTransform to PTransformProgress, to avoid overloading the name.) While we're on this, I might reserve "Metrics" to be the user metrics alone and put progress as a separate field in the response.

Reply:
The idea is that Metrics.PTransform would be what stored well known structured metrics while Metrics.User would store arbitrary unstructured user metrics.

This would allow Metrics.PTransform to add additional information beyond the scope of progress reporting since metrics could also include other information that isn't really progress related but could be health related such as memory consumed, CPU utilization, ...

Reply:
Well, I would call them "user metrics" and "system metrics" or something like that. As "user metrics" are also per PTransform, calling them ptransform metrics is confusing, and there's a move to make user metrics more structured as well. Happy to defer until user metrics are actually defined.

Reply:
Yes, we could reorganize the structure of Metrics so that User falls under PTransform or something else but we can defer that till a future time.